### ISSUE TITLE
pack_to_max_length = False后卡死

### ISSUE THREAD

**Issue by bo-jpg** [InternLM/xtuner#818]
pack_to_max_length = True 可以正常运行，pack_to_max_length = False后卡死，log打到"Checkpoints will be saved to...”就卡住了，也不报错，请问是怎么回事啊？

**Comment by HIT-cwh** [comment]
方便提供下您的训练config吗

**Comment by bo-jpg** [comment]
> 很抱歉给您的使用带来不便。为了我们进一步定位问题，麻烦提供一下训练 config 以及启动脚本

十分感谢回复！下面是config、启动脚本和环境：

**config**：

# Copyright (c) OpenMMLab. All rights reserved.
"""Data format:

[{
    "messages": [
        { "role": "system", "content": "xxx." },
        { "role": "user", "content": "xxx." },
        { "role": "assistant", "content": "xxx.", "loss": false},
        { "role": "user", "content": "xxx." },
        { "role": "assistant", "content": "xxx.", "loss": true}
    ]
},
...
]
"""  # noqa: E501
import torch
from datasets import load_dataset
from mmengine.config import read_base
from mmengine.dataset import DefaultSampler
from xtuner.dataset.samplers import LengthGroupedSampler
from mmengine.hooks import (CheckpointHook, DistSamplerSeedHook, IterTimerHook,
                            LoggerHook, ParamSchedulerHook)
from mmengine.optim import AmpOptimWrapper, CosineAnnealingLR, LinearLR
from peft import LoraConfig
from torch.optim import AdamW
from transformers import (AutoModelForCausalLM, AutoTokenizer,
                          BitsAndBytesConfig)

from xtuner.dataset import process_hf_dataset
from xtuner.dataset.collate_fns import default_collate_fn
from xtuner.dataset.map_fns import template_map_fn_factory
from xtuner.engine.hooks import (DatasetInfoHook, EvaluateChatHook,
                                 VarlenAttnArgsToMessageHubHook)
from xtuner.engine.runner import TrainLoop
from xtuner.model import SupervisedFinetune
from xtuner.utils import PROMPT_TEMPLATE

with read_base():
    from .map_f import single_turn_map_f as dataset_map_fn

#######################################################################
#                          PART 1  Settings                           #
#######################################################################
# Model

pretrained_model_name_or_path = "/home/model/Qwen2-7B"

use_varlen_attn = False

# Data

data_files = ["/home/data/train.json"]

prompt_template = PROMPT_TEMPLATE.qwen_chat

max_length = 100000

pack_to_max_length = False

# Scheduler & Optimizer
batch_size = 1  # per_device
accumulative_counts = 6  # bs = 1 GPU * 1 batch_size_per_device * 16 acc
dataloader_num_workers = 0
max_epochs = 10
optim_type = AdamW
lr = 5e-6
betas = (0.9, 0.999)
weight_decay = 0
max_norm = 1  # grad clip
warmup_ratio = 0.03

# Save

save_steps = 100
save_total_limit = -1  # Maximum checkpoints to keep (-1 means unlimited)

# Evaluate the generation performance during the training
evaluation_freq = 500
SYSTEM = ''
evaluation_inputs = [
    '请给我介绍五个上海的景点', 'Please tell me five scenic spots in Shanghai'
]

sequence_parallel_size = 8

#######################################################################
#                      PART 2  Model & Tokenizer                      #
#######################################################################
tokenizer = dict(
    type=AutoTokenizer.from_pretrained,
    pretrained_model_name_or_path=pretrained_model_name_or_path,
    trust_remote_code=True,
    padding_side='right',
    eos_token='<|im_end|>')


model = dict(
    type=SupervisedFinetune,
    use_varlen_attn=use_varlen_attn,
    llm=dict(
        type=AutoModelForCausalLM.from_pretrained,
        pretrained_model_name_or_path=pretrained_model_name_or_path,
        trust_remote_code=True,
        torch_dtype=torch.float16))

#######################################################################
#                      PART 3  Dataset & Dataloader                   #
#######################################################################
train_dataset = dict(
    type=process_hf_dataset,
    dataset=dict(type=load_dataset, path='json', data_files=data_files),
    tokenizer=tokenizer,
    max_length=max_length,
    dataset_map_fn=dataset_map_fn,
    template_map_fn=dict(
        type=template_map_fn_factory, template=prompt_template),
    remove_unused_columns=True,
    shuffle_before_pack=True,
    pack_to_max_length=pack_to_max_length,
    use_varlen_attn=use_varlen_attn)

train_dataloader = dict(
    batch_size=batch_size,
    num_workers=dataloader_num_workers,
    dataset=train_dataset,
    sampler=dict(type=DefaultSampler, shuffle=True),
    collate_fn=dict(type=default_collate_fn, use_varlen_attn=use_varlen_attn))

#######################################################################
#                    PART 4  Scheduler & Optimizer                    #
#######################################################################
# optimizer
optim_wrapper = dict(
    type=AmpOptimWrapper,
    optimizer=dict(
        type=optim_type, lr=lr, betas=betas, weight_decay=weight_decay),
    clip_grad=dict(max_norm=max_norm, error_if_nonfinite=False),
    accumulative_counts=accumulative_counts,
    loss_scale='dynamic',
    dtype='float16')

# learning policy
# More information: https://github.com/open-mmlab/mmengine/blob/main/docs/en/tutorials/param_scheduler.md  # noqa: E501
param_scheduler = [
    dict(
        type=LinearLR,
        start_factor=1e-5,
        by_epoch=True,
        begin=0,
        end=warmup_ratio * max_epochs,
        convert_to_iter_based=True),
    dict(
        type=CosineAnnealingLR,
        eta_min=0.0,
        by_epoch=True,
        begin=warmup_ratio * max_epochs,
        end=max_epochs,
        convert_to_iter_based=True)
]

# train, val, test setting
train_cfg = dict(type=TrainLoop, max_epochs=max_epochs)

#######################################################################
#                           PART 5  Runtime                           #
#######################################################################
# Log the dialogue periodically during the training process, optional
custom_hooks = [
    dict(type=DatasetInfoHook, tokenizer=tokenizer),
    dict(
        type=EvaluateChatHook,
        tokenizer=tokenizer,
        every_n_iters=evaluation_freq,
        evaluation_inputs=evaluation_inputs,
        system=SYSTEM,
        prompt_template=prompt_template)
]

if use_varlen_attn:
    custom_hooks += [dict(type=VarlenAttnArgsToMessageHubHook)]

# configure default hooks
default_hooks = dict(
    # record the time of every iteration.
    timer=dict(type=IterTimerHook),
    # print log every 10 iterations.
    logger=dict(type=LoggerHook, log_metric_by_epoch=False, interval=10),
    # enable the parameter scheduler.
    param_scheduler=dict(type=ParamSchedulerHook),
    # save checkpoint per `save_steps`.
    checkpoint=dict(
        type=CheckpointHook,
        by_epoch=False,
        interval=save_steps,
        max_keep_ckpts=save_total_limit),
    # set sampler seed in distributed evrionment.
    sampler_seed=dict(type=DistSamplerSeedHook),
)

# configure environment
env_cfg = dict(
    # whether to enable cudnn benchmark
    cudnn_benchmark=False,
    # set multi process parameters
    mp_cfg=dict(mp_start_method='fork', opencv_num_threads=0),
    # set distributed parameters
    dist_cfg=dict(backend='nccl'),
)

# set visualizer
visualizer = None

# set log level
log_level = 'INFO'

# load from which checkpoint
load_from = None

# whether to resume training from the loaded checkpoint
resume = False

# Defaults to use random seed and disable `deterministic`
randomness = dict(seed=None, deterministic=False)

# set log processor
log_processor = dict(by_epoch=False)


**启动脚本**：

GPU_NUM=8
CONFIG_PATH="/home/code/qwen_7b_custom_sft_copy.py"
OUT_DIR=/home/result/Qwen2-sft
NPROC_PER_NODE=${GPU_NUM} xtuner train ${CONFIG_PATH} --work-dir ${OUT_DIR} --deepspeed deepspeed_zero2


**环境**：
torch==2.1.2
cuda==12.1
transfomers==4.41.2
flash-attn==2.5.9post1
xtuner==0.1.21
deepspeed==0.14.0

**Comment by AEProgrammer** [comment]
一样的问题，使用pack_to_max_length=False卡死，不报错也不动

**Comment by bo-jpg** [comment]
@HIT-cwh @AEProgrammer 我使用最新版0.1.22的xtuner，这个问题仍然存在，不知道为啥

**Comment by HIT-cwh** [comment]
@bo-jpg 您这里开了序列并行，需要对应修改sampler为 SequenceParallelSampler ，类似这样 https://github.com/InternLM/xtuner/blob/main/xtuner/configs/llama/llama3_8b/llama3_8b_full_alpaca_e3.py#L93C11-L94

**Comment by bo-jpg** [comment]
> @bo-jpg 您这里开了序列并行，需要对应修改sampler为 SequenceParallelSampler ，类似这样 https://github.com/InternLM/xtuner/blob/main/xtuner/configs/llama/llama3_8b/llama3_8b_full_alpaca_e3.py#L93C11-L94

使用SequenceParallelSampler 可以了，感谢！另外可以顺带解释一下SequenceParallelSampler 的采样方式吗，比起DefaultSampler和LengthGroupedSampler有什么不同？

**Comment by HIT-cwh** [comment]
开启了序列并行之后，dp size = world size // sp size，相同dp_rank不同sp_rank的gpu，需要采样相同的数据，forward前需要对数据沿着seq_len维度做切分。

**Comment by bo-jpg** [comment]
@HIT-cwh 了解了，感谢！

另外，我的配置：
use_varlen_attn = False
pack_to_max_length = False
batch_size = 1  # per_device
accumulative_counts = 6  # bs = 1 GPU * 1 batch_size_per_device * 16 acc
max_epochs = 10
我的训练数据有3551条，但是log打印：
07/22 14:34:08 - mmengine - INFO - Num train samples 3427，以及
07/22 14:34:42 - mmengine - INFO - Iter(train) [   10/34270]  lr: 4.3867e-08  eta: 1 day, 4:36:17  time: 3.0058  data_time: 0.0179  memory: 36687  loss: 0.7363
07/22 14:35:12 - mmengine - INFO - Iter(train) [   20/34270]  lr: 9.2552e-08  eta: 1 day, 4:31:40  time: 2.9913  data_time: 0.0314  memory: 49394  loss: 0.6501

请问：1.3551为什么变成了3427？2.为什么总iter数是34270，怎么算的？
辛苦解答！
